elastic/elastic-package
# [system testing] Implement assertion for expected number of events

## What does this PR do?

This PR will enhance `system testing` through the implementation of an assertion for expected number of events. 

## Checklist

- [x] unpack events embedded in a message/data/events/etc. field within a hit
- [x] add test case to `httpjson` integration package
- [x] ensure test does not fail when test parameters are absent
- ~[ ] handle persistence between calls~ _NOT YET REQUIRED_
- ~[ ] sum events across calls~ _IMPLICIT_

## Author's Checklist

- [x] run system tests vs elasticsearch + 4 randomly selected integrations to check for regressions
    - [x] 1password
    - ~[ ] cisco_aironet~
    - [x] elasticsearch
    - [x] hadoop
    - ~[ ] juniper_junos~
    - ~[ ] kafka_log~
    - [x] nginx
    - [x] traefik
    - ~[ ] zookeeper~
- [x] add test under `other` 

## How to test this PR locally

Pull this branch and install elastic-package with `make install`

Clone `elastic/integrations`

Configure the httpjson (or related) integration as follows

For an http request, consider this example from the `httpjson/generic` data stream's `test-expected-hit-count-config.yml`, shown below.

```
input: httpjson
service: httpjson
data_stream:
  vars:
    data_stream.dataset: httpjson.generic
    username: test
    password: test
    request_url: http://{{Hostname}}:{{Port}}/testexpectedhits/api
    response_split: |- 
      target: body.hits
      type: array
      keep_parent: false     
assert:
  hit_count: 4
```

The `data_stream.vars.request_url` corresponds to a test-stub path in the `_dev/deploy/docker/files/config.yml` file.

```
  - path: /testexpectedhits/api
    methods: ["GET"]
    request_headers:
      Authorization:
        - "Basic dGVzdDp0ZXN0"
    responses:
      - status_code: 200
        headers:
          Content-Type:
            - "application/json; charset=utf-8"
        body: |-
          {"total":3,"hits":[{"message": "success"},{"message": "success"},{"message": "success"}]}
```

As `assert.hit_count` is defined and `> 0` the system test will only pass the assertion when the number of elements in the top-level array of the JSON body is equal to this value. In this case the assertion requires 4 hits when the actual count will be 3, so the assertion should fail.

Run the system tests using `elastic-package test system -v`

The test results should include the following line:

```
│ httpjson │ generic     │ system    │ expected-hit-count │ FAIL: observed hit count 3 did not match expected hit count 4 │ 21.123980175s │
```

Next, amend `assert.hit_count` so that this matches the number of elements (3) in the response to the `/testexpectedevents/api` path. The test should now pass as follows

```
│ httpjson │ generic     │ system    │ expected-hit-count │ PASS   │ 20.979543465s │
```

## Related issues

- Relates https://github.com/elastic/elastic-package/issues/821

~## Screenshots~
